x/concentrated-liquidity/lp.go

```diff
@@ -17,88 +17,152 @@ import (
 // Returns error if:
 // TODO: list error cases
 // TODO: table-driven tests
-func (k Keeper) createPosition(ctx sdk.Context, poolId uint64, owner sdk.AccAddress, amount0Desired, amount1Desired, amount0Min, amount1Min sdk.Int, lowerTick, upperTick int64) (amtDenom0, amtDenom1 sdk.Int, liquidityCreated sdk.Dec, err error) {
-	// ensure types.MinTick <= lowerTick < types.MaxTick
-	// TODO (bez): Add unit tests.
-	if lowerTick < types.MinTick || lowerTick >= types.MaxTick {
-		return sdk.Int{}, sdk.Int{}, sdk.Dec{}, fmt.Errorf("invalid lower tick: %d", lowerTick)
+func (k Keeper) createPosition(ctx sdk.Context, poolId uint64, owner sdk.AccAddress, amount0Desired, amount1Desired, amount0Min, amount1Min sdk.Int, lowerTick, upperTick int64) (sdk.Int, sdk.Int, sdk.Dec, error) {
+	if err := validateTickRangeIsValid(lowerTick, upperTick); err != nil {
+		return sdk.Int{}, sdk.Int{}, sdk.Dec{}, err
 	}
 
-	// ensure types.MaxTick < upperTick <= types.MinTick
-	// TODO (bez): Add unit tests.
-	if upperTick > types.MaxTick || upperTick <= types.MinTick {
-		return sdk.Int{}, sdk.Int{}, sdk.Dec{}, fmt.Errorf("invalid upper tick: %d", upperTick)
+	sqrtPriceLowerTick, sqrtPriceUpperTick, err := ticksToSqrtPrice(lowerTick, upperTick)
+	if err != nil {
+		return sdk.Int{}, sdk.Int{}, sdk.Dec{}, err
 	}
 
 	// now calculate amount for token0 and token1
-	pool := k.getPoolbyId(ctx, poolId)
-
-	currentSqrtPrice := pool.CurrentSqrtPrice
-	sqrtRatioUpperTick, err := tickToSqrtPrice(sdk.NewInt(upperTick))
+	pool, err := k.getPoolbyId(ctx, poolId)
 	if err != nil {
 		return sdk.Int{}, sdk.Int{}, sdk.Dec{}, err
 	}
-	sqrtRatioLowerTick, err := tickToSqrtPrice(sdk.NewInt(lowerTick))
+
+	liquidityDelta := getLiquidityFromAmounts(pool.CurrentSqrtPrice, sqrtPriceLowerTick, sqrtPriceUpperTick, amount0Desired, amount1Desired)
+	if liquidityDelta.IsZero() {
+		return sdk.Int{}, sdk.Int{}, sdk.Dec{}, fmt.Errorf("liquidity delta zero")
+	}
+
+	actualAmount0, actualAmount1, err := k.updatePosition(ctx, poolId, owner, lowerTick, upperTick, liquidityDelta)
 	if err != nil {
 		return sdk.Int{}, sdk.Int{}, sdk.Dec{}, err
 	}
 
-	liquidity := getLiquidityFromAmounts(currentSqrtPrice, sqrtRatioLowerTick, sqrtRatioUpperTick, amount0Desired, amount1Desired)
-	if liquidity.IsZero() {
-		return sdk.Int{}, sdk.Int{}, sdk.Dec{}, fmt.Errorf("token in amount is zero")
+	// TODO: handle amount0Min, amount1Min
```

Comment:
can we just add the check here now to ensure actual0 and actual1 are greater than mins?

Reply:
done